dimod/quadratic/cyqm/cyqm_template.pyx.pxi

```diff
@@ -185,16 +193,38 @@ cdef class cyQM_template(cyQMBase):
             for vi in range(length):
                 self.add_linear(vi, linear[vi])
 
-    def add_quadratic(self, u, v, bias_type bias):
-        cdef Py_ssize_t ui = self.variables.index(u)
-        cdef Py_ssize_t vi = self.variables.index(v)
-
-        if ui == vi and (self.cppqm.vartype(ui) == cppVartype.SPIN
-                         or self.cppqm.vartype(ui) == cppVartype.BINARY):
-            raise ValueError(f"{u!r} cannot have an interaction with itself")
+    cdef Py_ssize_t _add_quadratic(self, index_type ui, index_type vi, bias_type bias) except -1:
+        # note: does not test that ui, vi are valid
+
+        if ui == vi:
+            if self.cppqm.vartype(ui) == cppVartype.SPIN:
+                raise ValueError(f"SPIN variables (e.g. {self.variables[ui]!r}) "
+                                 "cannot have interactions with themselves"
+                                 )
+            if self.cppqm.vartype(vi) == cppVartype.BINARY:
+                raise ValueError(f"BINARY variables (e.g. {self.variables[ui]!r}) "
+                                 "cannot have interactions with themselves"
+                                 )
+
+        if not self.REAL_INTERACTIONS:
+            if self.cppqm.vartype(ui) == cppVartype.REAL:
+                raise ValueError(
+                    f"REAL variables (e.g. {self.variables[ui]!r}) "
+                    "cannot have interactions with other variables"
+                    )
```

Comment:
Shouldn't this be ``"cannot have interactions")`` ? It seems it will complain of an interaction between REAL and REAL too, right?